vault/sealunwrapper.go

```diff
@@ -0,0 +1,183 @@
+	if entry == nil {
+		return nil
+	}
+
+	locksutil.LockForKey(d.locks, entry.Key).Lock()
+	defer locksutil.LockForKey(d.locks, entry.Key).Unlock()
+
+	return d.underlying.Put(ctx, entry)
+}
+
+func (d *sealUnwrapper) Get(ctx context.Context, key string) (*physical.Entry, error) {
+	entry, err := d.underlying.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var performUnwrap bool
+	se := &physical.SealWrapEntry{}
+	// If the value ends in our canary value, try to decode the bytes.
+	eLen := len(entry.Value)
+	if eLen > 0 && entry.Value[eLen-1] == 's' {
+		if err := proto.Unmarshal(entry.Value[:eLen-1], se); err == nil {
+			// We unmarshaled successfully which means we need to store it as a
+			// non-proto message
+			performUnwrap = true
+		}
+	}
+	if !performUnwrap {
+		return entry, nil
+	}
+	// It's actually encrypted and we can't read it
+	if se.Wrapped {
+		return nil, fmt.Errorf("cannot decode sealwrapped storage entry %s", entry.Key)
+	}
+	if atomic.LoadUint32(d.allowUnwraps) != 1 {
+		return &physical.Entry{
+			Key:   entry.Key,
+			Value: se.Ciphertext,
+		}, nil
+	}
+
+	locksutil.LockForKey(d.locks, key).Lock()
+	defer locksutil.LockForKey(d.locks, key).Unlock()
+
+	// At this point we need to re-read and re-check
+	entry, err = d.underlying.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	performUnwrap = false
+	se = &physical.SealWrapEntry{}
+	// If the value ends in our canary value, try to decode the bytes.
+	eLen = len(entry.Value)
+	if eLen > 0 && entry.Value[eLen-1] == 's' {
+		// We ignore an error because the canary is not a guarantee; if it
+		// doesn't decode, proceed normally
+		if err := proto.Unmarshal(entry.Value[:eLen-1], se); err == nil {
+			// We unmarshaled successfully which means we need to store it as a
+			// non-proto message
+			performUnwrap = true
+		}
+	}
+	if !performUnwrap {
+		return entry, nil
+	}
+	if se.Wrapped {
+		return nil, fmt.Errorf("cannot decode sealwrapped storage entry %s", entry.Key)
+	}
+
+	entry = &physical.Entry{
+		Key:   entry.Key,
+		Value: se.Ciphertext,
+	}
+
+	if atomic.LoadUint32(d.allowUnwraps) != 1 {
+		return entry, nil
+	}
+	return entry, d.underlying.Put(ctx, entry)
+}
+
+func (d *sealUnwrapper) Delete(ctx context.Context, key string) error {
+	locksutil.LockForKey(d.locks, key).Lock()
+	defer locksutil.LockForKey(d.locks, key).Unlock()
+
+	return d.underlying.Delete(ctx, key)
+}
+
+func (d *sealUnwrapper) List(ctx context.Context, prefix string) ([]string, error) {
+	return d.underlying.List(ctx, prefix)
+}
+
+func (d *transactionalSealUnwrapper) Transaction(ctx context.Context, txns []*physical.TxnEntry) error {
+	// Collect keys that need to be locked
+	var keys []string
+	for _, curr := range txns {
+		keys = append(keys, curr.Entry.Key)
+	}
+	// Lock the keys
+	for _, l := range locksutil.LocksForKeys(d.locks, keys) {
+		l.Lock()
```

Comment:
Does this not cause a deadlock when 2 keys return the same lock?

Reply:
LockForKeys is relatively new and will generate a list of unique locks.